File: data_structures/trees/trie/python/prefix_tree.py
from typing import Union, Dict, List
from data_structures.stack.pyhton.stack import Stack


class Trie:
    val: str = None
    children: Dict[str, 'Trie']
    is_word: bool = False

    def __init__(self, val: Union[str, None] = None):
        self.val = val
        self.children = {}

    def insert(self, word: str) -> None:
        curr = self
        for char in word:
            curr.children[char] = curr.children[char] if char in curr.children else Trie(char)
            curr = curr.children[char]

        curr.is_word = True

    def search(self, word: str, partial: bool = False) -> bool:
        curr = self.search_node(word)

        if isinstance(curr, Trie):
            return True if partial else curr.is_word

        return False

    def search_node(self, word: str) -> Union['Trie', bool]:
        curr = self

        for char in word:
            if char not in curr.children:
                return False
            curr = curr.children[char]

        return curr

    def starts_with(self, prefix: str) -> bool:
        return self.search(prefix, True)

    def get_all_words(self,prefix: str = '', node: 'Trie' = None) -> List[str]:
        if node is None:
            node = self

        words: List[str] = []
        if not node:
            return words

        if node.is_word:
            words.append(prefix)

        for char in node.children.keys():
            new_words = self.get_all_words(f'{prefix}{char}', node.children[char])
            words = words + new_words

        return words

    def autocomplete(self, prefix: str = '') -> List[str]:
        curr = self.search_node(prefix)
        start_node = curr if isinstance(curr, Trie) else False

        if not start_node:
            return []

        return self.get_all_words(prefix, start_node)

    def remove(self, word: str) -> bool:
        curr = self
        stack = Stack[Trie]()

        for char in word:
            if char not in curr.children:
                return False

            curr = curr.children[char]
            stack.push(curr)

        child: Trie = stack.pop()
        child.is_word = False

        while not stack.is_empty:
            parent: Trie = stack.pop()
            if not child.is_word and not len(child.children.keys()):
                del parent.children[child.val]
            child = parent

        return True

File: data_structures/trees/trie/typescript/PrefixTree.ts
/**
Example representation
Trie {
    val: 'o'
    children: {
        w: Trie {
            val: 'w'
            children: {
                l: Trie {
                    val: 'l'
                    children: {}
                    isWord = true
                }
            }
            isWord = false
        }
    }
    isWord = false
}
**/
import Stack from '@ds/stack/typescript/Stack';

export default class Trie {
    val: string;
    children: Record<string, this | Trie> = {};
    isWord = false;

    constructor(val: string | undefined = undefined) {
        this.val = val;
    }

    insert(word: string): void {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        let curr: this | Trie = this;
        for (const char of word) {
            curr.children[char] = curr.children[char] || new Trie(char);
            curr = curr.children[char];
        }
        curr.isWord = true;
    }

    search(word: string, partial = false): boolean {
        const curr = this.searchNode(word);

        if (curr instanceof Trie) {
            return partial ? true : curr.isWord;
        }

        return false;
    }

    searchNode(word: string): Trie | boolean {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        let curr: this | Trie = this;

        for (const char of word) {
            if (!curr.children[char]) {
                return false;
            }
            curr = curr.children[char];
        }

        return curr;
    }

    remove(word: string): boolean {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        let curr: this | Trie = this;
        const stack = new Stack<Trie>();

        for (const char of word) {
            if (!curr.children[char]) {
                return false;
            }

            curr = curr.children[char];
            stack.push(curr);
        }

        let child = stack.pop();
        child.isWord = false;

        // remove non-words without children
        while (!stack.isEmpty) {
            const parent = stack.pop();
            if (!child.isWord && !Object.keys(child.children).length) {
                delete parent.children[child.val];
            }
            child = parent;
        }

        return true;
    }

    startsWith(prefix: string): boolean {
        return this.search(prefix, true);
    }

    getAllWords(prefix = '', node: Trie = this): string[] {
        let words: string[] = [];

        if (node.isWord) {
            words.push(prefix);
        }

        for (const char of Object.keys(node.children)) {
            const newWords = this.getAllWords(`${prefix}${char}`, node.children[char]);
            words = words.concat(newWords);
        }

        return words;
    }

    autoComplete(prefix = ''): string[] {
        const curr = this.searchNode(prefix);
        const startNode = curr instanceof Trie ? curr : false;

        if (!startNode) {
            return [];
        }

        return this.getAllWords(prefix, startNode);
    }
}
